src/Adapter/Tornado/EventLoop.php

```diff
@@ -97,11 +107,26 @@ function (\Throwable $throwable) use ($task) {
      */
     public function async(\Generator $generator): Promise
     {
-        $this->tasks[] = ($task = new Internal\Task(
-            $generator,
-            Internal\PendingPromise::createUnhandled($this->unhandledFailingPromises))
+        /** @var Task $task */
+        $task = null;
+        /** @var PendingPromise $promise */
+        $promise = null;
+
+        $promise = Internal\PendingPromise::createUnhandled($this->unhandledFailingPromises,
+            function () use (&$task, &$promise) {
+                if ($task->getGenerator()->current()) {
+                    $promise->reject(new CancelledException('async cancellation'));
+                    /** @var PendingPromise $currentPromise */
+                    $currentPromise = $task->getGenerator()->current();
+                    if (!$currentPromise->isCancelled()) {
```

Comment:
This test seems useless since `PendingPromise::cancel` already check this (and it is its responsibility!)